File: sdks/dotnet/sulfone-helium/Domain/Core/CyanScript.cs
using sulfone_helium.Domain.Core.FileSystem;
using sulfone_helium.Domain.Plugin;
using sulfone_helium.Domain.Processor;
using sulfone_helium.Domain.Resolver;

namespace sulfone_helium.Domain.Core;

public interface ICyanTemplate
{
    Task<Cyan> Template(IInquirer inquirer, IDeterminism determinism);
}

public interface ICyanProcessor
{
    Task<ProcessorOutput> Process(CyanProcessorInput input, CyanFileHelper fileHelper);
}

public interface ICyanPlugin
{
    Task<PluginOutput> Plugin(CyanPluginInput input);
}

public interface ICyanResolver
{
    Task<ResolverOutput> Resolve(ResolverInput input);
}

File: sdks/python/cyan_sdk/domain/core/cyan_script.py
from abc import abstractmethod, ABC

from cyan_sdk.domain.core.cyan import Cyan
from cyan_sdk.domain.core.cyan_script_model import CyanExtensionInput, CyanProcessorInput, CyanPluginInput
from cyan_sdk.domain.core.deterministic import IDeterminism
from cyan_sdk.domain.core.fs.cyan_fs_helper import CyanFileHelper
from cyan_sdk.domain.core.inquirer import IInquirer
from cyan_sdk.domain.plugin.output import PluginOutput
from cyan_sdk.domain.processor.output import ProcessorOutput


class ICyanTemplate(ABC):
    @abstractmethod
    async def template(self, inquirer: IInquirer, determinism: IDeterminism) -> Cyan:
        pass


class ICyanExtension(ABC):
    @abstractmethod
    async def extension(self, inquirer: IInquirer, determinism: IDeterminism, i: CyanExtensionInput) -> Cyan:
        pass


class ICyanProcessor(ABC):
    @abstractmethod
    async def process(self, i: CyanProcessorInput, file_helper: CyanFileHelper) -> ProcessorOutput:
        pass


class ICyanPlugin(ABC):
    @abstractmethod
    async def plugin(self, i: CyanPluginInput) -> PluginOutput:
        pass
